[libc][bazel] Prevent LIBC_NAMESPACE leakage

gchatelet opened the pull request:
Right now `LIBC_NAMESPACE` definition is added to all rules that depend upon
any libc function. This have far-reaching effects which prevents llvm-libc
development when the installed libc is itself llvm-libc.

This PR removes the `:libc_root` target entirely and consistently use the
provided libc BUILD actions for compilation and testing. This makes sure that we
don't leak `LIBC_NAMESPACE` to our client.

gchatelet committed:
Remove special handling of :libc_root in libc_function

gchatelet committed:
Update bazel config

michaelrj-google commented:
I like the move to every target being through a `libc_*` instead of just `cc_*`. 
I'm not confident enough in my bazel to comment on the design of the rest of the change, but it works fine in my testing.

gchatelet committed:
Do not prefix label with @llvm-project
This helps downstream transformations
Also inline function - it's not used elsewhere.

gchatelet committed:
Do not hide symbols in libc_support_library

gchatelet committed:
Use :libc instead of //libc

gchatelet commented:
Thx for the review @michaelrj-google . Can you PTAL?

lntue commented on utils/bazel/llvm-project-overlay/libc/libc_build_rules.bzl:
```diff
@@ -48,7 +48,7 @@ def _libc_library(name, hidden, copts = [], deps = [], **kwargs):
 # libc_support_library.
 def libc_support_library(name, **kwargs):
-    _libc_library(name = name, hidden = True, **kwargs)
+    _libc_library(name = name, hidden = False, **kwargs)
```
Why are we not hiding support library symbols anymore?

gchatelet replied:
Hiding the symbols results in internal fuzzers complaining about different visibility between modules:
```
default visibility for functions and variables [-fvisibility] differs in PCH file vs. current file
module file __redacted__/stl/_objs/stl_cc_library/stl_cc_library.pic.pcm cannot be loaded due to a configuration mismatch with the current compilation [-Wmodule-file-config-mismatch]
```

`hidden = False` is actually coherent with how it worked before this patch. I wanted to straighten up the visibility but it has consequences for downstream.

Now that I think about it, it makes sense that both internal targets and libc_support_library are compiled with the same visibility. I'll fix it by using `libc_support_library` in lieu of `_libc_library` in `libc_function` below.

gchatelet replied:
@lntue does it look good to you?

michaelrj-google reviewed (approved):
Patch works on my end, please address Tue's comment before landing.

gchatelet committed:
Make sure internal targets are compiled with the same visibility as support libraries
